src/queue.rs

```diff
@@ -0,0 +1,102 @@
+//! The queue of nodes that keeps track of pending operations.
+
+use crate::node::Node;
+use crate::sync::atomic::{AtomicPtr, Ordering};
+
+use crossbeam_utils::CachePadded;
+
+use alloc::boxed::Box;
+use core::ptr;
+
+/// A queue of nodes.
+pub(crate) struct Queue {
```

Comment:
why can't we use `concurrent-queue` for this?

Reply:
I used `concurrent-queue` in #30 and it led to a significant slowdown compared to the current implementation. Although, the new queue isn't on the hot path anymore so I doubt it matters that much.

Also, `concurrent-queue` hasn't released a version yet with the `no_std` patch. I can file a PR for that if you want.

Reply:
file a PR please, I would really want to avoid code duplication/diverging optimizations between multiple smoi-rs crates.

Reply:
Here it is: smol-rs/concurrent-queue#29

Reply:
What is the difference between the algorithms used in this queue impl vs. the one in `concurrent-queue`? That is, are here any improvements contained which could be upstreamed into `concurrent-queue`?

Reply:
The one used in this queue impl is a fairly naive algorithm that allocates per item. The main advantage of this algorithm over something like `concurrent-queue` is the fact that, in the best case scenario, there are fewer atomic operations. I would not recommend upstreaming this `concurrent-queue`.

Reply:
With version 2.0.0 of `concurrent-queue`, I've updated this PR to use `concurrent-queue`.